File: trunk/Demos/Crazy Stars/MenuItem.cpp
#include "StdAfx.h"
#include "MenuItem.h"

CMenuItem::CMenuItem(int nID, HFONTX hFont, HAUDIO hSound, float x, float y, float fDelay, LPSTR lpcszTitle)
	: CControl(nID)
	, m_hFont(hFont)
	, m_hSound(hSound)
	, m_fDelay(fDelay)
	, m_lpcszTitle(lpcszTitle)
	, m_fOffsetSrc(0)
	, m_fOffsetDelta(0)
	, m_fOffset(0)
	, m_fTime(-1)
	, m_fTime2(-1)
{
	m_colColor.SetColor(0xFFFFE060);
	m_colShadow.SetColor(0x30000000);

	m_bStatic=FALSE;
	m_bVisible=TRUE;
	m_bEnabled=TRUE;

	RECT rect={0, 0, 0, 0};

	m_pApp->Font_DrawText(m_hFont, m_lpcszTitle, &rect, 0, DT_CALCRECT);
	m_Rect.Set(x-(float)rect.right/2, y, x+(float)rect.right/2, y+rect.bottom);
}

CMenuItem::~CMenuItem()
{
}

void CMenuItem::Render()
{
	RECT rect={(LONG)m_Rect.x1+(LONG)m_fOffset+3, (LONG)m_Rect.y1+3, 0, 0};

	m_pApp->Font_DrawText(m_hFont, m_lpcszTitle, &rect, 0, DT_CALCRECT);
	m_pApp->Font_DrawText(m_hFont, m_lpcszTitle, &rect, 0, DT_TOP|DT_LEFT, m_colShadow.GetColor());
	rect.left=(LONG)m_Rect.x1-(LONG)m_fOffset;
	rect.top=(LONG)m_Rect.y1-(LONG)m_fOffset;
	m_pApp->Font_DrawText(m_hFont, m_lpcszTitle, &rect, 0, DT_CALCRECT);
	m_pApp->Font_DrawText(m_hFont, m_lpcszTitle, &rect, 0, DT_TOP|DT_LEFT, m_colColor.GetColor());
}

void CMenuItem::Update(float fDelta)
{
	if (m_fTime2!=-1)
	{
		m_fTime2+=fDelta;
		if (m_fTime2>=m_fDelay+0.1f)
		{
			m_colColor=m_colSrc2+m_colDelta2;
			m_colShadow=m_colShadowSrc+m_colShadowDelta;
			m_fOffset=0;
			m_fTime2=-1;
		}
		else
		{
			if (m_fTime2<m_fDelay)
			{
				m_colColor=m_colSrc2;
				m_colShadow=m_colShadowSrc;
			}
			else
			{
				m_colColor=m_colSrc2+m_colDelta2*(m_fTime2-m_fDelay)*10;
				m_colShadow=m_colShadowSrc+m_colShadowDelta*(m_fTime2-m_fDelay)*10;
			}
		}
	}
	else if (m_fTime!=-1)
	{
		m_fTime+=fDelta;
		if (m_fTime>=0.2f)
		{
			m_colColor=m_colSrc+m_colDelta;
			m_fOffset=m_fOffsetSrc+m_fOffsetDelta;
			m_fTime=-1.0f;
		}
		else
		{
			m_colColor=m_colSrc+m_colDelta*m_fTime*5;
			m_fOffset=m_fOffsetSrc+m_fOffsetDelta*m_fTime*5;
		}
	}
}

void CMenuItem::Enter()
{
	CColor colTemp;

	m_colSrc2.SetColor(0x00FFE060);
	colTemp.SetColor(0xFFFFE060);
	m_colDelta2=colTemp-m_colSrc2;
	m_colShadowSrc.SetColor(0x00000000);
	colTemp.SetColor(0x30000000);
	m_colShadowDelta=colTemp-m_colShadowSrc;
	m_fTime2=0.0f;
}

void CMenuItem::Exit()
{
	CColor colTemp;

	m_colSrc2.SetColor(0xFFFFE060);
	colTemp.SetColor(0x00FFE060);
	m_colDelta2=colTemp-m_colSrc2;
	m_colShadowSrc.SetColor(0x30000000);
	colTemp.SetColor(0x00000000);
	m_colShadowDelta=colTemp-m_colShadowSrc;
	m_fTime2=0.0f;
}

BOOL CMenuItem::IsDone()
{
	if (m_fTime2==-1.0f) return TRUE;
	else return FALSE;
}

void CMenuItem::Focus(BOOL bFocused)
{
	CColor colTemp;
	
	if(bFocused)
	{
		m_pApp->Effect_Play(m_hSound);
		m_colSrc.SetColor(0xFFFFE060);
		colTemp.SetColor(0xFFFFFFFF);
		m_fOffsetSrc=0;
		m_fOffsetDelta=4;
	}
	else
	{
		m_colSrc.SetColor(0xFFFFFFFF);
		colTemp.SetColor(0xFFFFE060);
		m_fOffsetSrc=4;
		m_fOffsetDelta=-4;
	}
	m_colDelta=colTemp-m_colSrc;
	m_fTime=0.0f;
}

void CMenuItem::MouseOver(BOOL bOver)
{
	if (bOver) m_pGUI->SetFocus(m_nID);
}

BOOL CMenuItem::MouseLBtn(BOOL bDown)
{
	if (bDown)
	{
		m_pApp->Effect_Play(m_hSound);
		m_fOffset=0;
		return FALSE;
	}
	else
	{
		m_fOffset=4;
		return TRUE;
	}
}

BOOL CMenuItem::KeyClick(int nVKey, char chChar)
{
	if (nVKey==VK_RETURN)
	{
		MouseLBtn(TRUE);
		return MouseLBtn(FALSE);
	}
	return FALSE;
}

File: GameMaker/Core/Menu Item.lua
function Construct(this)
	OffsetSrc=0;
	OffsetDelta=0;
	Offset=0;
	Time=-1;
	Time2=-1;

	colColor=Color_SetColor(0xFFFFE060);
	colShadow=Color_SetColor(0x30000000);

	Control_SetStatic(this, false);
	Control_SetVisible(this, true);
	Control_SetEnabled(this, true);

	local n=0;
	local Rect=
	{
		Left=0,
		Top=0,
		Right=0,
		Bottom=0
	};

	n, Rect=Font_DrawText(Attribute.Font, Attribute.Title, Rect, 0, DT_CALCRECT);
	Rect.Left=Attribute.x-Rect.Right/2;
	Rect.Top=Attribute.y;
	Rect.Right=Attribute.x+Rect.Right/2;
	Rect.Bottom=Attribute.y+Rect.Bottom;
	Control_SetRect(this, Rect);
end

function Deconstruct(this)
end

function SetColor(this, Color)
end

function Render(this)
	local n=0;
	local mRect=Control_GetRect(this);
	local Rect=
	{
		Left=math.floor(mRect.Left+Offset+3),
		Top=math.floor(mRect.Top+3),
		Right=0,
		Bottom=0
	};

	n, Rect=Font_DrawText(Attribute.Font, Attribute.Title, Rect, 0, DT_CALCRECT);
	Font_DrawText(Attribute.Font, Attribute.Title, Rect, 0, Bit_or(DT_TOP, DT_LEFT), Color_GetColor(colShadow));
	Rect.Left=math.floor(mRect.Left-Offset);
	Rect.Top=math.floor(mRect.Top-Offset);
	n, Rect=Font_DrawText(Attribute.Font, Attribute.Title, Rect, 0, DT_CALCRECT);
	Font_DrawText(Attribute.Font, Attribute.Title, Rect, 0, Bit_or(DT_TOP, DT_LEFT), Color_GetColor(colColor));
end

function Update(this, Delta)
	if Time2~=-1 then
		Time2=Time2+Delta;
		if Time2>=Attribute.Delay+0.1 then
			colColor=Color_Add(colSrc2, colDelta2);
			colShadow=Color_Add(colShadowSrc, colShadowDelta);
			Offset=0;
			Time2=-1;
		else
			if Time2<Attribute.Delay then
				colColor=
				{
					a=colSrc2.a,
					r=colSrc2.r,
					g=colSrc2.g,
					b=colSrc2.b
				};
				colShadow=
				{
					a=colShadowSrc.a,
					r=colShadowSrc.r,
					g=colShadowSrc.g,
					b=colShadowSrc.b
				};
			else
				colColor=Color_Add(colSrc2, Color_Mul(colDelta2, (Time2-Attribute.Delay)*10));
				colShadow=Color_Add(colShadowSrc, Color_Mul(colShadowDelta, (Time2-Attribute.Delay)*10));
			end
		end
	elseif Time~=-1 then
		Time=Time+Delta;
		if Time>0.2 then
			colColor=Color_Add(colSrc, colDelta);
			Offset=OffsetSrc+OffsetDelta;
			Time=-1;
		else
			colColor=Color_Add(colSrc, Color_Mul(colDelta, Time*5));
			Offset=OffsetSrc+OffsetDelta*Time*5;
		end
	end
end

function Enter(this)
	colSrc2=Color_SetColor(0x00FFE060);
	local col=Color_SetColor(0xFFFFE060);
	colDelta2=Color_Sub(col, colSrc2);
	colShadowSrc=Color_SetColor(0);
	col=Color_SetColor(0x30000000);
	colShadowDelta=Color_Sub(col, colShadowSrc);
	Time2=0;
end

function Exit(this)
	colSrc2=Color_SetColor(0xFFFFE060);
	local col=Color_SetColor(0x00FFE060);
	colDelta2=Color_Sub(col, colSrc2);
	colShadowSrc=Color_SetColor(0x30000000);
	col=Color_SetColor(0);
	colShadowDelta=Color_Sub(col, colShadowSrc);
	Time2=0;
end

function Reset(this)
end

function IsDone(this)
	if Time2==-1 then return true;
	else return false; end
end

function Focus(this, bFocused)
	local col;

	if bFocused then
		Effect_Play(Attribute.Sound);
		colSrc=Color_SetColor(0xFFFFE060)
		col=Color_SetColor(0xFFFFFFFF);
		OffsetSrc=0;
		OffsetDelta=4;
	else
		colSrc=Color_SetColor(0xFFFFFFFF)
		col=Color_SetColor(0xFFFFE060);
		OffsetSrc=4;
		OffsetDelta=-4;
	end
	colDelta=Color_Sub(col, colSrc);
	Time=0;
end

function MouseOver(this, bOver)
	if bOver then GUI_SetFocus(Control_GetGUI(this), Control_GetID(this)); end
end

function MouseMove(this, x, y)
	return false;
end

function MouseLBtn(this, bDown, x, y)
	if bDown then
		Effect_Play(Attribute.Sound);
		Offset=0;
		return false;
	else
		Offset=4;
		return true;
	end
end

function MouseRBtn(this, bDown, x, y)
	return false;
end

function MouseWheel(this, Wheel)
	return false;
end

function KeyClick(this, VKey, Char)
	if VKey==VK_RETURN then
		MouseLBtn(this, true, 0, 0)
		return MouseLBtn(this, false, 0, 0);
	end
	return false;
end
